api/internal/localizer/localizer.go

```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
 	}
 	return locPath, nil
 }
+
+// localizeDir localizes root path and returns the localized path
+func (lc *Localizer) localizeDir(path string) (string, error) {
+	ldr, err := lc.ldr.New(path)
+	if err != nil {
+		return "", errors.Wrap(err)
+	}
+	defer func() { _ = ldr.Cleanup() }()
+
+	var locPath string
+	if repo := ldr.Repo(); repo != "" {
+		// TODO(annasong): check if able to add localize directory
+		locPath = locRootPath(path, filesys.ConfirmedDir(repo), filesys.ConfirmedDir(ldr.Root()))
+	} else {
+		locPath, err = filepath.Rel(lc.ldr.Root(), ldr.Root())
+		if err != nil {
+			// This should never occur, but if hit, something is fundamentally wrong and we should immediately exit.
```

Comment:
I think the main difference between using `log.Fatalf` and `return nil, err` is that the latter actually gives users more info to triage and fix (from the layered wrapped error messages). Do you mind further explaining the benefits of using `log.Fatalf` here? If it's not that obvious, maybe we shall return the error.  And if we return error, maybe we do not need to skip the gocritic linter (which is another bonus)?

Reply:
My initial reasoning for using `log.Fatal` here instead of `error` is to differentiate "impossible" errors - errors that should never be thrown - from normal errors. A `filepath.Rel` error would contradict our assumptions about `fileloader` checking for legal, navigable paths. If this error is thrown, there is a bug in our code that we should catch during testing or review. We should fail fast and there shouldn't be a way to silence/bypass this failure.

I believe there are similar use cases in kustomize code. For example, https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/filesys/fsondisk.go#L69

Down to change it to return the error if that's the better practice. Happy to defer to @yuwenma to make the call.

Reply:
That's fair.

Reply:
The counterargument is more theoretical than anything. The distinction between `log.Fatal` and `log.Panic` is that `log.Panic` allows for recovery and `defer` statements to run. In our use case, if this `log.Fatal` is hit, core assumptions have been violated, and we should exit immediately, no exceptions. We don't want to recover or give a chance for `defer` statements to further corrupt the existing state. The argument against needing to call `defer` statements to cleanup `localize` state is that because we only call `log.Fatal` under "impossible" cases, we should catch these cases during testing and shouldn't hit these cases during normal execution.

Given that this argument is purely theoretical, I can change instances of `log.Fatal` to `log.Panic` if you'd prefer, @KnVerey

Reply:
I see what you're saying about these being things that should never happen in normal execution, but in short, I prefer panicking in such cases (if there's no way around them to begin with!). I don't think there's much risk of us letting in code that recovers arbitrary panics (the only recovery I see is inside the go-yaml fork fwiw), and I feel like the cleanup code we typically have deferred should be safe to run regardless, e.g. I'd see cleaning up temporary state we've created on disk as desirable. Two other things to consider: first, if you allow the test execution to recover, you can have a nicer debugging experience where test cases other than the triggering case can continue. Second, if the problem _does_ end up happening in normal execution, the panic output is maximally clear that this is not the end user's fault and should be reported.

Reply:
Thanks for the detailed explanation!! Got it. Will change.